File: src/ios/SleepFocusPlugin.m
#import "SleepFocusPlugin.h"
#import <DeviceActivity/DeviceActivity.h>

@implementation SleepFocusPlugin

- (void)checkSleepFocus:(CDVInvokedUrlCommand*)command {
    if (@available(iOS 15.0, *)) {
        if ([FocusStatusCenter.shared.authorizationStatus isEqualToNumber:@(FocusAuthorizationStatusAuthorized)]) {
            BOOL sleepFocusActive = [[FocusStatusCenter.shared.focusStatuses filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(id obj, NSDictionary *bindings) {
                return [obj isFocusedForActivity:FocusActivitySleep];
            }]] count] > 0;

            NSDictionary* result = @{@"status": sleepFocusActive ? @"active" : @"inactive"};
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:result];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } else {
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Authorization not granted"];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }
    } else {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"iOS version too low"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

@end

File: src/ios/SleepFocusPlugin.swift
import Foundation
import DeviceActivity

@objc(SleepFocusPlugin)
public class SleepFocusPlugin: CDVPlugin {
    @objc public func checkSleepFocus(_ command: CDVInvokedUrlCommand) {
        if #available(iOS 15.0, *) {
            let authorizationStatus = FocusStatusCenter.shared.authorizationStatus
            if authorizationStatus == .authorized {
                let sleepFocusActive = FocusStatusCenter.shared.focusStatuses.contains { $0.isFocused(for: .sleep) }
                let result: [String: Any] = ["status": sleepFocusActive ? "active" : "inactive"]
                self.commandDelegate.send(CDVPluginResult(status: .ok, messageAs: result), callbackId: command.callbackId)
            } else {
                self.commandDelegate.send(CDVPluginResult(status: .error, messageAs: "Authorization not granted"), callbackId: command.callbackId)
            }
        } else {
            self.commandDelegate.send(CDVPluginResult(status: .error, messageAs: "iOS version too low"), callbackId: command.callbackId)
        }
    }
}
